cmd/haproxy-cfg-generator/cfg_generator.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2020 Authors of Arktos.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this tsFile except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+var (
+	templateFile        string
+	cfgGenerated        string
+	tenantPartititonIPs []string
+	resourcePartitionIP string
+)
+
+const (
+	ARKTOS_API_PORT      = 8080
+	MAX_CONN_PER_BACKEND = 500000
+	CONNECTION_TIMEOUT   = "10m"
+	PROXY_PORT           = "8888"
+)
+
+func initFlags() {
+	flag.StringVar(&templateFile, "template", "", "The path to the haproxy cfg file template")
+	flag.StringVar(&cfgGenerated, "target", "", "The path to the generated haproxy cfg file")
+
+	flag.Parse()
+
+	validateFlags()
+}
+
+func validateFlags() {
+	if strings.TrimSpace(cfgGenerated) == "" {
+		LogError("\nNo target file is specified.\n")
+		os.Exit(1)
+	}
+
+	if !FileExists(templateFile) {
+		LogError("\nTemplate file (%v) does not exist.\n", templateFile)
+		os.Exit(1)
+	}
+}
+
+func getPartitionIPs() {
+	TP_IP_str := strings.TrimSpace(os.Getenv("TENANT_PARTITION_IP"))
+	if len(TP_IP_str) == 0 {
+		LogError("\nTENANT_PARTITION_IP is not specified.\n")
+		os.Exit(1)
+	}
+
+	for _, ip := range strings.Split(TP_IP_str, ",") {
+		ip_addr := strings.TrimSpace(ip)
+		if net.ParseIP(ip_addr) == nil {
+			LogError("\nIP address (%v) is invalid.\n", ip_addr)
+			os.Exit(1)
+		}
+		tenantPartititonIPs = append(tenantPartititonIPs, ip_addr)
+	}
+
+	if len(tenantPartititonIPs) > 26 {
+		LogError("\nUnable to support TP # > 26.\n")
+		os.Exit(1)
+	}
+
+	resourcePartitionIP = strings.TrimSpace(os.Getenv("RESOURCE_PARTITION_IP"))
+	if len(resourcePartitionIP) == 0 {
+		LogError("\nRESOURCE_PARTITION_IP is not specified.\n")
+		os.Exit(1)
+	}
+
+	if net.ParseIP(resourcePartitionIP) == nil {
+		LogError("\nIP address (%v) is invalid.\n", resourcePartitionIP)
+		os.Exit(1)
+	}
+}
+
+func get_tp_request_acl() string {
```

Comment:
some functions in this file with "fancy" logic might need unit tests

Reply:
As discussed, this tool is now test-protected.

Also the "patch haproxy for prometheus" will be activated only when a specific env var is "true".